File: Leetcode/3583 - Count Special Triplets/solution.ts
function specialTriplets(nums: number[]): number {
    const MOD: number = 1e9 + 7;

    const totalIndicesLessThan = (indices: number[], target: number): number => {
        let start: number = 0;
        let end: number = indices.length;
        while (start < end) {
            const mid: number = start + Math.floor((end - start) / 2);
            if (indices[mid] < target) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return start;
    };

    const totalIndicesGreaterThan = (indices: number[], target: number): number => {
        let start: number = 0;
        let end: number = indices.length - 1;
        let resIndex: number = indices.length;
        while (start <= end) {
            const mid: number = start + Math.floor((end - start) / 2);
            if (indices[mid] > target) {
                resIndex = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return indices.length - resIndex;
    };

    let numsMap: Record<number, number[]> = {};
    for (let i: number = 0; i < nums.length; i++) {
        const num: number = nums[i];
        if (!numsMap[num]) numsMap[num] = [];
        numsMap[num].push(i);
    }

    let count: number = 0;
    for (const jNumStr in numsMap) {
        const jNum: number = parseInt(jNumStr);
        const ikNum: number = jNum * 2;

        if (!numsMap[ikNum]) continue;
        for (let j: number = 0; j < numsMap[jNum].length; j++) {
            const jIndex: number = numsMap[jNum][j];
            const iTotalIndices: number = totalIndicesLessThan(numsMap[ikNum], jIndex);
            const kTotalIndices: number = totalIndicesGreaterThan(numsMap[ikNum], jIndex);
            count = (count + (iTotalIndices * kTotalIndices) % MOD) % MOD;
        }
    }

    return count;
};

File: Leetcode/3583 - Count Special Triplets/solution.py
def specialTriplets(self, nums: List[int]) -> int:
    MOD: int = 10**9 + 7

    def totalIndicesLessThan(indices: List[int], target: int) -> int:
        start: int = 0
        end: int = len(indices)
        while start < end:
            mid: int = start + (end - start) // 2
            if indices[mid] < target:
                start = mid + 1
            else:
                end = mid
        return start

    def totalIndicesGreaterThan(indices: List[int], target: int) -> int:
        start: int = 0
        end: int = len(indices) - 1
        resIndex: int = len(indices)
        while start <= end:
            mid: int = start + (end - start) // 2
            if indices[mid] > target:
                resIndex = mid
                end = mid - 1
            else:
                start = mid + 1
        return len(indices) - resIndex

    numsMap: Dict[int, List[int]] = {}

    for i in range(len(nums)):
        num: int = nums[i]
        if num not in numsMap:
            numsMap[num] = []
        numsMap[num].append(i)

    for index_list in numsMap.values():
        index_list.sort()

    count: int = 0

    for jNum in numsMap:
        ikNum: int = jNum * 2
        if ikNum not in numsMap:
            continue

        for jIndex in numsMap[jNum]:
            iTotalIndices: int = totalIndicesLessThan(numsMap[ikNum], jIndex)
            kTotalIndices: int = totalIndicesGreaterThan(numsMap[ikNum], jIndex)
            count = (count + (iTotalIndices * kTotalIndices) % MOD) % MOD

    return count